src/style-spec/function/expression.js

```diff
@@ -0,0 +1,253 @@
+        return new ParsingContext(
+            this.definitions,
+            path,
+            expectedType || null,
+            scope,
+            this.errors
+        );
+    }
+
+    /**
+     * Push a parsing (or type checking) error into the `this.errors`
+     * @param error The message
+     * @param keys Optionally specify the source of the error at a child
+     * of the current expression at `this.key`.
+     * @private
+     */
+    error(error: string, ...keys: Array<number>) {
+        const key = `${this.key}${keys.map(k => `[${k}]`).join('')}`;
+        this.errors.push(new ParsingError(key, error));
+        return null;
+    }
+}
+
+/**
+ * Parse the given JSON expression.
+ *
+ * @param expectedType If provided, the parsed expression will be checked
+ * against this type.  Additionally, `expectedType` will be pssed to
+ * Expression#parse(), wherein it may be used to infer child expression types
+ *
+ * @private
+ */
+function parseExpression(expr: mixed, context: ParsingContext) : ?Expression {
+    if (expr === null || typeof expr === 'string' || typeof expr === 'boolean' || typeof expr === 'number') {
+        expr = ['literal', expr];
+    }
+
+    if (Array.isArray(expr)) {
+        if (expr.length === 0) {
+            return context.error(`Expected an array with at least one element. If you wanted a literal array, use ["literal", []].`);
+        }
+
+        const op = expr[0];
+        if (typeof op !== 'string') {
+            context.error(`Expression name must be a string, but found ${typeof op} instead. If you wanted a literal array, use ["literal", [...]].`, 0);
+            return null;
+        }
+
+        const Expr = context.definitions[op];
+        if (Expr) {
+            const parsed = Expr.parse(expr, context);
+            if (!parsed) return null;
+            if (context.expectedType && match(context.expectedType, parsed.type, context)) {
+                return null;
+            } else {
+                return parsed;
+            }
+        }
+
+        return context.error(`Unknown expression "${op}". If you wanted a literal array, use ["literal", [...]].`, 0);
+    } else if (typeof expr === 'undefined') {
+        return context.error(`'undefined' value invalid. Use null instead.`);
+    } else if (typeof expr === 'object') {
+        return context.error(`Bare objects invalid. Use ["literal", {...}] instead.`);
+    } else {
+        return context.error(`Expected an array, but found ${typeof expr} instead.`);
+    }
+}
+
+/**
+ * Returns null if the type matches, or an error message if not.
+ *
+ * If `context` is provided, then also push the error to it via
+ * `context.error()`
+ *
+ * @private
+ */
+function match(
+    expected: Type,
+    t: Type,
+    context?: ParsingContext
+): ?string {
+    let error = `Expected ${expected.name} but found ${t.name} instead.`;
+
+    // a `null` literal is allowed anywhere.
```

Comment:
Yeah, the case you spelled out is the one I had in mind.  Hard to be anything like sure, but my guess is that this will be relatively common, because datasets with missing or poorly typed property values are quite common.  That said, with the `let` option allowing us to avoid writing the fallback expression more than once, I'm okay with our just recommending `[let, o, O, [coalesce, [match, input, ..., O], O]` -- we can always revisit if this turns out to be common and too inconvenient.